cluster/assemblestate/transport.go

```diff
@@ -0,0 +1,371 @@
+	return &HTTPTransport{
+		logger: logger,
+	}
+}
+
+func (h *HTTPTransport) Serve(ctx context.Context, addr string, cert tls.Certificate, as *AssembleState) error {
+	mux := http.NewServeMux()
+	mux.Handle("/assemble/auth", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		h.handleAuth(w, r, as)
+	}))
+	mux.Handle("/assemble/routes", h.trustedHandler(h.handleRoutes, as))
+	mux.Handle("/assemble/unknown", h.trustedHandler(h.handleUnknown, as))
+	mux.Handle("/assemble/devices", h.trustedHandler(h.handleDevices, as))
+
+	server := &http.Server{
+		Handler: mux,
+	}
+
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		listener := tls.NewListener(ln, &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth:   tls.RequireAnyClientCert,
+		})
+
+		_ = server.Serve(listener)
+	}()
+
+	// wait for context cancellation
+	<-ctx.Done()
+
+	// shutdown the server
+	_ = server.Shutdown(context.Background())
+	wg.Wait()
+
+	return nil
+}
+
+type countingReader struct {
+	r     io.Reader
+	count int64
+}
+
+func (c *countingReader) Read(p []byte) (int, error) {
+	n, err := c.r.Read(p)
+	c.count += int64(n)
+	return n, err
+}
+
+func (h *HTTPTransport) trustedHandler(next func(http.ResponseWriter, *http.Request, *PeerHandle), as *AssembleState) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.TLS == nil {
+			w.WriteHeader(403)
+			return
+		}
+
+		if len(r.TLS.PeerCertificates) != 1 {
+			w.WriteHeader(403)
+			return
+		}
+
+		cert := r.TLS.PeerCertificates[0].Raw
+		peer, err := as.VerifyPeer(cert)
+		if err != nil {
+			h.logger.Debug("dropping message from untrusted peer")
+			w.WriteHeader(403)
+			return
+		}
+
+		next(w, r, peer)
+	}
+}
+
+func (h *HTTPTransport) handleAuth(w http.ResponseWriter, r *http.Request, as *AssembleState) {
+	if r.Method != "POST" {
+		w.WriteHeader(405)
+		return
+	}
+
+	if r.TLS == nil {
+		w.WriteHeader(400)
+		return
+	}
+
+	if len(r.TLS.PeerCertificates) != 1 {
+		w.WriteHeader(400)
+		return
+	}
+
+	// set a max size so an untrusted peer can't send some massive JSON
+	const maxAuthSize = 1024 * 4
+	counter := countingReader{
+		r: http.MaxBytesReader(w, r.Body, maxAuthSize),
+	}
+
+	var auth Auth
+	if err := json.NewDecoder(&counter).Decode(&auth); err != nil {
+		w.WriteHeader(400)
+		return
+	}
+
+	cert := r.TLS.PeerCertificates[0].Raw
+	if err := as.Authenticate(auth, cert); err != nil {
+		w.WriteHeader(403)
+		return
+	}
+
+	h.received.Add(counter.count)
+}
+
+func (h *HTTPTransport) handleRoutes(w http.ResponseWriter, r *http.Request, peer *PeerHandle) {
+	if r.Method != "POST" {
+		w.WriteHeader(405)
+		return
+	}
+
+	counter := countingReader{r: r.Body}
+
+	var routes Routes
+	if err := json.NewDecoder(&counter).Decode(&routes); err != nil {
+		w.WriteHeader(400)
+		return
+	}
+
+	err := peer.AddRoutes(routes)
```

Comment:
Add is a strange verb here, we are taking routes from the peer, maybe Accept/Take ...? same for other cases here